lib/runtime/src/transports/etcd/connector.rs

```diff
@@ -107,3 +113,57 @@ impl Connector {
         &self.connect_options
     }
 }
+
+#[derive(Debug)]
+struct BackoffState {
+    /// Initial backoff duration for reconnection attempts
+    pub initial_backoff: Duration,
+    /// Maximum backoff duration for reconnection attempts
+    pub max_backoff: Duration,
+    /// Current backoff duration (starts at 0 for immediate reconnect)
+    current_backoff: Duration,
+    /// Last time a connection establishment was attempted
+    last_connect_attempt: std::time::Instant,
+}
+
+impl Default for BackoffState {
+    fn default() -> Self {
+        Self {
+            initial_backoff: Duration::from_millis(500),
+            max_backoff: Duration::from_secs(5),
+            current_backoff: Duration::ZERO,
+            last_connect_attempt: std::time::Instant::now(),
+        }
+    }
+}
+
+impl BackoffState {
+    /// Reset backoff to 0 if enough time has passed since the last connection
+    pub fn attempt_reset(&mut self) {
+        if std::time::Instant::now() > self.last_connect_attempt + self.current_backoff {
+            tracing::debug!("Resetting backoff to 0 (first reconnect or enough time has passed)");
+            self.current_backoff = Duration::ZERO;
+        }
+    }
+
+    /// Apply backoff and update backoff state for possible next connection attempt
+    pub async fn apply_backoff(&mut self, remaining: Duration) {
+        if self.current_backoff > Duration::ZERO {
+            let backoff = std::cmp::min(
+                std::cmp::min(self.current_backoff, remaining / 2),
+                self.max_backoff,
+            );
```

Comment:
Do we want to `max(1, <val>)` here, so we don't get a backoff of 0? The ` / 2` will round to 0 at some point.